go/runtime/txpool/check_queue.go

```diff
@@ -1,155 +1,85 @@
 package txpool
 
 import (
-	"container/list"
 	"fmt"
 	"sync"
 
-	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
-	"github.com/oasisprotocol/oasis-core/go/runtime/host/protocol"
+	"github.com/gammazero/deque"
 )
 
-type pendingTx struct {
-	Tx       []byte
-	TxHash   hash.Hash
-	Meta     *TransactionMeta
-	NotifyCh chan *protocol.CheckTxResult
-
-	element *list.Element
-}
-
-// checkTxQueue is a queue backed by an ordered map.
 type checkTxQueue struct {
-	sync.Mutex
+	l sync.Mutex
 
-	transactions map[hash.Hash]*pendingTx
-	queue        *list.List
+	txs *deque.Deque
 
-	maxTxPoolSize uint64
-	maxBatchSize  uint64
+	maxSize      int
+	maxBatchSize int
 }
 
-// Add adds transaction into the queue.
-func (q *checkTxQueue) Add(tx *pendingTx) error {
-	q.Lock()
-	defer q.Unlock()
+func (cq *checkTxQueue) add(pct *PendingCheckTransaction) error {
+	cq.l.Lock()
+	defer cq.l.Unlock()
 
 	// Check if there is room in the queue.
-	if uint64(q.queue.Len()) >= q.maxTxPoolSize {
-		return fmt.Errorf("tx pool is full")
-	}
-
-	if err := q.checkTxLocked(tx.Tx, tx.TxHash); err != nil {
-		return err
+	if cq.txs.Len() >= cq.maxSize {
+		return fmt.Errorf("check queue is full")
 	}
 
-	q.addTxLocked(tx)
+	cq.txs.PushBack(pct)
 
 	return nil
 }
 
-// GetBatch gets a batch of transactions from the queue.
-func (q *checkTxQueue) GetBatch() []*pendingTx {
-	q.Lock()
-	defer q.Unlock()
+func (cq *checkTxQueue) retryBatch(pcts []*PendingCheckTransaction) {
+	cq.l.Lock()
+	defer cq.l.Unlock()
+
+	// NOTE: This is meant for retries so it ignores the size limit on purpose.
```

Comment:
Is there something in place to prevent unbounded growth?  I guess with the accounting in add, the queue size will be at most `2 *cq.maxSize`?

Reply:
So this is only used for reinserting previously poped transactions (which are at most "batch size" long). In the worst case this could overflow the queue to be twice the max size (in case of check batches being that big -- which they are not by default) when the queue is full and batch checks are failing while new transactions are being queued.